File: gh_stuff/gh_combat_deck.py
"""Combat deck simulator
"""

import numpy

CONDITION_MUDDLE = 1 
CONDITION_POISON = 2 
CONDITION_WOUND = 3 
CONDITION_IMMOBILIZE = 4 
CONDITION_STUN = 5 
CONDITION_ADD_TARGET = 6 

class Condition(object):
    _CONDITION_MAP = {
        CONDITION_MUDDLE: "muddle",
        CONDITION_POISON: "poison",
        CONDITION_WOUND: "wound", 
        CONDITION_IMMOBILIZE: "immobilize", 
        CONDITION_STUN: "stun",  
        CONDITION_ADD_TARGET: "target-one-more", 
    }

    def __init__(self, condition):
        self._condition = condition
        return

    def __str__(self):
        return self._CONDITION_MAP[self._condition]

class CritModifier(object):
    def __str__(self):
        return "2x"

class NullModifier(object):
    def __str__(self):
        return "~null~"

class SpecialBlessing(object):
    def __str__(self):
        return "BLESSING"
    
class SpecialCurse(object):
    def __str__(self):
        return "CURSE"

class Card(object):
    def __init__(self, modifier=0, rolling=False, condition=None, reshuffle=False, special=None):
        self._modifier = modifier
        self._rolling = rolling
        self._condition = condition
        self._reshuffle = reshuffle
        self._special = special
        return

    def __str__(self):
        str_mod = "{:+d}".format(self._modifier) if isinstance(self._modifier, int) else self._modifier
        str_roll = "[rolling]" if self._rolling else ""
        str_cond = "[{}]".format(self._condition) if self._condition else ""
        str_reshuffle = "RESHUFFLE" if self._reshuffle else "" 
        str_special = "<{}>".format(self._special) if self._special else ""

        output = "    {mod} {roll} {cond} {reshuffle} {special}".format(
            mod=str_mod,
            roll=str_roll,
            cond=str_cond,
            reshuffle=str_reshuffle,
            special=str_special,
        )

        return output

    def __repr__(self):
        str_mod = "{:+d}".format(self._modifier) if isinstance(self._modifier, int) else self._modifier
        str_roll = "[rolling]" if self._rolling else ""
        str_cond = "[{}]".format(self._condition) if self._condition else ""
        str_reshuffle = "RESHUFFLE" if self._reshuffle else "" 
        str_special = "<{}>".format(self._special) if self._special else ""

        output = "<Card: {mod} {roll} {cond} {reshuffle} {special}>".format(
            mod=str_mod,
            roll=str_roll,
            cond=str_cond,
            reshuffle=str_reshuffle,
            special=str_special,
        )

        return output

    def __mul__(self, other):
        if not isinstance(other, int):
            raise TypeError("For multiplication, expected int; {} found".format(type(other).__name__))
        return CardStack([self] * other)

    def __rmul__(self, other):
        if not isinstance(other, int):
            raise TypeError("For multiplication, expected int; {} found".format(type(other).__name__))
        return CardStack(other * [self])

    def __add__(self, other):
        if isinstance(other, Card):
            return CardStack([self, other])
        elif isinstance(other, CardStack):
            other.insert(self)
            return other
        else:
            raise TypeError("For addition, expected Card or CardStack; {} found".format(type(other).__name__))

    @property
    def is_one_time_use(self):
        return isinstance(self._special, SpecialCurse) or isinstance(self._special, SpecialBlessing)
        

class CardStack(object):
    def __init__(self, cards=[]):
        if isinstance(cards, CardStack):
            self._cards = list(cards._cards)
            return
        
        self._cards = []

        for item in cards:
            if isinstance(item, Card):
                self.append(item)
            elif isinstance(item, CardStack):
                self += item
            else:
                raise TypeError("For addition, expected Card or CardStack; {} found".format(type(other).__name__))
                
        return

    def append(self, card):
        self._cards.append(card)
        return

    def insert(self, card):
        self._cards.insert(card, 0)
        return

    def __str__(self):
        cards_str = "\n    ".join([str(c) for c in self._cards])
        output = "CardStack\n    {}".format(cards_str)
        return output

    def __repr__(self):
        return "<CardStack with {} cards>".format(len(self._cards))

    def __len__(self):
        return len(self._cards)

    def __getitem__(self, idx):
        return self._cards[idx]

    def __add__(self, other):
        if isinstance(other, CardStack):
            s = CardStack()
            s._cards = self._cards + other._cards
            return s
        elif isinstance(other, Card):
            s = CardStack(self)
            s.append(other)
            return s
        else:
            raise TypeError("For addition, expected Card or CardStack; {} found".format(type(other).__name__))

    def __iadd__(self, other):
        if isinstance(other, Card):
            self.append(other)
        if isinstance(other, CardStack):
            self._cards += other._cards
        return self

    def __mul__(self, other):
        if not isinstance(other, int):
            raise TypeError("For multiplication, expected int; {} found".format(type(other).__name__))
        s = CardStack(self) 
        s._cards *= other
        return s

    def __rmul__(self, other):
        if not isinstance(other, int):
            raise TypeError("For multiplication, expected int; {} found".format(type(other).__name__))
        s = CardStack(self) 
        s._cards *= other
        return s

    def __imul__(self, other):
        if not isinstance(other, int):
            raise TypeError("For multiplication, expected int; {} found".format(type(other).__name__))
        self._cards *= other
        return self

    def __iter__(self):
        for c in self._cards:
            yield c


class Deck(object):
    CARDS_IN_DECK = CardStack([
        Card(modifier=NullModifier(), reshuffle=True), 
        Card(modifier=-2),
        Card(modifier=-1) * 5,
        Card(modifier=0) * 6,
        Card(modifier=1) * 5,
        Card(modifier=2),
        Card(modifier=CritModifier(), reshuffle=True),
    ])
    def __init__(self):
        self.reset()
        return

    def reset(self):
        self._cardstack = CardStack(self.CARDS_IN_DECK)
        self._to_be_removed = []
        self._playable_indices = list(range(len(self._cardstack)))
        self._k = 0

        self._shuffled_indices = numpy.array(self._playable_indices)
        self.shuffle()
        self._total = len(self._playable_indices)
        return

    def shuffle(self):
        for idx in self._to_be_removed:
            self._playable_indices.remove(idx)

        self._shuffled_indices = numpy.array(self._playable_indices)
        numpy.random.shuffle(self._shuffled_indices)
        self._total = len(self._playable_indices)
        self._k = 0
        return

    def undo(self):
        if self._k > 0:
            self._k -= 1
        return

    def deal(self):
        if self._k >= len(self._playable_indices):
            return None

        c = self._cardstack[self._shuffled_indices[self._k]]
        if c.is_one_time_use:
            self._to_be_removed.append(self._shuffled_indices[self._k])
        self._k += 1
        return c
    
    def add_blessing(self):
        self._cardstack.append(Card(modifier=CritModifier(), special=SpecialBlessing()))
        new_idx = len(self._cardstack) - 1
        self._playable_indices.append(new_idx)
        self._shuffled_indices = numpy.append(self._shuffled_indices[self._k:], new_idx)
        numpy.random.shuffle(self._shuffled_indices)
        return

    def add_curse(self):
        self._cardstack.append(Card(modifier=NullModifier(), special=SpecialCurse()))
        new_idx = len(self._cardstack) - 1
        self._playable_indices.append(new_idx)
        self._shuffled_indices = numpy.append(self._shuffled_indices[self._k:], new_idx)
        numpy.random.shuffle(self._shuffled_indices)
        return

    @property
    def total(self):
        return self._total

    @property
    def index(self):
        return self._k

    
class Repl(object):
    def __init__(self, deck=Deck()):
        self._deck = deck
        return

    def prompt(self):
        the_prompt = "{name} ({index:d}/{total:d}): ".format(
            name=self._deck.__class__.__name__,
            index=self._deck.index,
            total=self._deck.total,
        )
        return input(the_prompt)
    

File: gh_stuff/gh_combat_deck.js
function shuffle(arr) {
    for (let ii = arr.length - 1; ii > 0; ii--) {
        let jj = Math.floor(Math.random() * ii);
        let tmp = arr[ii];
        arr[ii] = arr[jj];
        arr[jj] = tmp;
    }
    return arr;
}

const ModifierType = {
    NEG_TWO: "-2",
    NEG_ONE: "-1",
    ZERO: "+0",
    ONE: "+1",
    TWO: "+2",
    CRIT: "2X",
    NULL: "~null~"
}
Object.freeze(ModifierType);

const ConditionType = {
    NONE: "",
    MUDDLE: "muddle",
    POISON: "poison",
    WOUND: "wound", 
    IMMOBILIZE: "immobilize",
    STUN: "stun",
    CURSE: "curse",
    ADD_TARGET: "target-one-more", 
}
Object.freeze(ConditionType);

function Condition(condition) {
    this.condition = condition;
}

Condition.prototype.toString = function conditionToString() {
    return `${this.condition}`;
}

const SpecialCardType = {
    NONE: "",
    BLESSING: "BLESSING",
    CURSE: "CURSE"
}
Object.freeze(SpecialCardType);

function SpecialCard(special) {
    this.special = special;
}

SpecialCard.prototype.toString = function specialToString() {
    return `${this.special}`;
}

function Card(modifier=ModifierType.ZERO,
              rolling=false,
              condition=ConditionType.NONE,
              reshuffle=false,
              special=SpecialCardType.NONE) {
    this.modifier = modifier;
    this.rolling = rolling;
    this.condition = condition;
    this.reshuffle = reshuffle;
    this.special = special;
}

Card.prototype.toString = function cardToString() {
    let num_str = `${this.modifier}`;
    let rolling_str = this.rolling ? "[rolling]" : "";
    let condition_str = this.condition ? `[${this.condition}]` : "";
    let reshuffle_str = this.reshuffle ? "RESHUFFLE" : "";
    let special_str = this.special ? `<${this.special}>` : "";
    return `    ${num_str} ${rolling_str} ${condition_str} ${reshuffle_str} ${special_str}`
}

// TODO: Convert this to toHtml()
//       Move display() functionality to Deck
Card.prototype.toHtml = function cardToHtml() {
    let num_str = `    ${this.modifier}`;
    let rolling_str = this.rolling ? " [rolling]" : "";
    let condition_str = this.condition ? ` [${this.condition}]` : "";
    let reshuffle_str = this.reshuffle ? " RESHUFFLE" : "";
    let special_str = this.special ? ` <${this.special}>` : "";


    var p = document.createElement("p");

    function add_card_attr(desc, cls) {
        mark = document.createElement("mark");
        mark.append(document.createTextNode(desc));
        mark.classList.add(cls);
        p.append(mark);
    }

    p.append(document.createTextNode("\u00A0\u00A0\u00A0\u00A0"));

    add_card_attr(num_str, "modifier");
    add_card_attr(rolling_str, "rolling");
    add_card_attr(condition_str, "condition");
    add_card_attr(reshuffle_str, "reshuffle");
    add_card_attr(special_str, "special");
    
    p.classList.add("card");
    return p;
}

Card.prototype.isOneTimeUse = function cardIsOneTimeUse() {
    return this.special ? true : false;
}

const BASIC_DECK = [
    // Basic combat deck
    {modifier: ModifierType.NULL, rolling: false, condition: ConditionType.NONE, reshuffle: true, special: SpecialCardType.NONE},
    {modifier: ModifierType.NEG_TWO, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    // {modifier: ModifierType.NEG_ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    // {modifier: ModifierType.NEG_ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    // {modifier: ModifierType.NEG_ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    // {modifier: ModifierType.NEG_ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.NEG_ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    // {modifier: ModifierType.ZERO, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    // {modifier: ModifierType.ZERO, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    // {modifier: ModifierType.ZERO, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    // {modifier: ModifierType.ZERO, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ZERO, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ZERO, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ONE, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.TWO, rolling: false, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.CRIT, rolling: false, condition: ConditionType.NONE, reshuffle: true, special: SpecialCardType.NONE},
    // Additions
    {modifier: ModifierType.ONE, rolling: true, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ONE, rolling: true, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ONE, rolling: true, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ONE, rolling: true, condition: ConditionType.NONE, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ZERO, rolling: true, condition: ConditionType.ADD_TARGET, reshuffle: false, special: SpecialCardType.NONE},
    {modifier: ModifierType.ZERO, rolling: false, condition: ConditionType.STUN, reshuffle: false, special: SpecialCardType.NONE},
]


function Deck(target_elem, card_configs=BASIC_DECK, cards_to_show=15) {
    this.target_elem = target_elem;
    this.card_configs = [...card_configs];
    this.playable_indices = [...Array(this.card_configs.length).keys()];
    this.to_be_removed = [];
    this.played = [];
    this.shuffled_indices = shuffle([...this.playable_indices]);
    this.k = 0;
    this.cards_to_show = cards_to_show;
    this.num_cards_in_play = this.playable_indices.length;
}

Deck.prototype.shuffle = function deckShuffle() {
    // Remove indices to cards that have been marked for removal
    for (let ii = 0; ii < this.to_be_removed.length; ii++) {
        this.playable_indices = this.playable_indices.filter(idx => idx !== this.to_be_removed[ii]);
    }

    this.shuffled_indices = shuffle([...this.playable_indices]);
    this.k = 0;
    this.num_cards_in_play = this.playable_indices.length;
};

Deck.prototype.undo = function deckUndo() {
    if (this.k > 0) {
        this.k--;
        this.target_elem.removeChild(this.target_elem.firstChild);
        if (this.played.length > this.cards_to_show) {
            this.target_elem.append(this.played[this.played.length - this.cards_to_show - 1].toHtml());
        }
        this.played.pop();
    }
};

Deck.prototype.deal = function deckDeal() {
    let conf = this.card_configs[this.shuffled_indices[this.k]];
    if (!conf) {
        var p = document.createElement("p");
        p.append(document.createTextNode("E M P T Y\u00A0\u00A0\u00A0\u00A0D E C K !"));
        this.target_elem.prepend(p);
        this.k++;
        return;
    }

    let c = new Card(conf.modifier, conf.rolling, conf.condition, conf.reshuffle, conf.special);
    if (c.isOneTimeUse()) {
        this.to_be_removed.push(this.shuffled_indices[this.k]);
    }

    let card_elem = c.toHtml();
    mark = document.createElement("mark");
    mark.append(document.createTextNode(`${this.k + 1}/${this.num_cards_in_play}:\u00A0`));
    mark.classList.add("counter");
    card_elem.prepend(mark);
    
    this.target_elem.prepend(card_elem);

    // while (this.target_elem.childElementCount > this.cards_to_show) {
    //     this.target_elem.removeChild(this.target_elem.lastChild);
    // }

    this.k++;
    this.played.push(c);
};

Deck.prototype.addBlessing = function deckAddBlessing() {
    this.card_configs.push({
        modifier: ModifierType.CRIT,
        rolling: false,
        condition: ConditionType.NONE,
        reshuffle: false,
        special: SpecialCardType.BLESSING
    })

    let new_idx = this.card_configs.length - 1;
    this.playable_indices.push(new_idx);
    this.shuffled_indices.push(new_idx);
    this.shuffled_indices.splice(this.k,
                                 this.num_cards_in_play - this.k + 1,
                                 ...shuffle(this.shuffled_indices.slice(this.k)));
};

Deck.prototype.addCurse = function deckAddCurse() {
    this.card_configs.push({
        modifier: ModifierType.NULL,
        rolling: false,
        condition: ConditionType.NONE,
        reshuffle: false,
        special: SpecialCardType.CURSE
    })

    let new_idx = this.card_configs.length - 1;
    this.playable_indices.push(new_idx);
    this.shuffled_indices.push(new_idx);
    this.shuffled_indices.splice(this.k,
                                 this.num_cards_in_play - this.k + 1,
                                 ...shuffle(this.shuffled_indices.slice(this.k)));
};

Deck.prototype.show = function deckShow(card) {
    if (card) {
        this.target_elem.prepend(card.toHtml());
    } else {
        var p = document.createElement("p");
        p.append(document.createTextNode("E M P T Y\u00A0\u00A0\u00A0\u00A0D E C K !"));
        this.target_elem.prepend(p);
    }

    while (this.target_elem.childElementCount > this.cards_to_show) {
        this.target_elem.removeChild(this.target_elem.lastChild);
    }
};

Deck.prototype.clearView = function deckClearView() {
    this.target_elem.innerHTML = "";
};

function addButtonListener(elem, listener) {
    if (elem.addEventListener) {
        elem.addEventListener("click", listener, false);
    } else if (elem.attachEvent) {
        elem.attachEvent("onclick", listener);
    }
}

// TODOs:
//    * [done] Implement a card deck
//    * Implement user interaction
//    * Implement config saving and loading


function displayData() {}

let db;

window.onload = function() {
    let request = window.indexedDB.open("foobar_db", 1);

    request.onerror = function() {
        console.log('Database failed to open.');
    };

    request.onsuccess = function() {
        console.log('Database opened successfully.');
        db = request.result;
        displayData();
    };

    request.onupgradeneeded = function(ev) {
        let db = ev.target.result;
        let objectStore = db.createObjectStore("foobar_os", {keyPath: 'id', autoIncrement: true});
        objectStore.createIndex('title', 'title', {unique: false});
        objectStore.createIndex('body', 'body', {unique: false});

        console.log('Database setup complete');
    };
}

function addData(title, body) {
    let newItem = {title: title, body: body};
    let transaction = db.transaction(['foobar_os'], 'readwrite');
    let objectStore = transaction.objectStore('foobar_os');
    let request = objectStore.add(newItem);

    request.onsuccess = function() {
        console.log('data added successfully...');
    };

    transaction.oncomplete = function() {
        console.log('Transaction complete.  Database modification finished.');
        displayData();
    };

    transaction.onerror = function() {
        console.log('Transaction failed to open.');
    };

    
}
